lib/vector-core/src/stream/batcher.rs

```diff
@@ -0,0 +1,540 @@
+    /// in this structure.
+    allocation_limit: usize,
+    /// The store of `I` elements.
+    elements: Vec<I>,
+}
+
+impl<I> ByteSizeOf for Batch<I> {
+    fn allocated_bytes(&self) -> usize {
+        self.allocated_bytes
+    }
+}
+
+impl<I> Batch<I>
+where
+    I: ByteSizeOf,
+{
+    /// Create a new Batch instance
+    ///
+    /// Creates a new batch instance with specific element and allocation
+    /// limits. The element limit is a maximum cap on the number of `I`
+    /// instances. The allocation limit is a soft-max on the number of allocated
+    /// bytes stored in this batch, not taking into account overhead from this
+    /// structure itself. Caller is responsible for ensuring that `I` will fit
+    /// inside the allocation limit.
+    ///
+    /// # Panics
+    ///
+    /// This function will panic if the allocation limit will not store at least
+    /// 1 instance of `I`, as measured by `mem::size_of`.
+    fn new(element_limit: usize, allocation_limit: usize) -> Self {
+        assert!(allocation_limit >= mem::size_of::<I>());
+        Self {
+            allocated_bytes: 0,
+            element_limit,
+            allocation_limit,
+            elements: Vec::with_capacity(element_limit),
+        }
+    }
+
+    /// Unconditionally insert an element into the batch
+    ///
+    /// This function is similar to `push` except that the caller does not need
+    /// to call `has_space` prior to calling this and it will never
+    /// panic. Intended to be used only when insertion must not fail.
+    fn with(mut self, value: I) -> Self {
+        self.allocated_bytes += value.size_of();
+        self.elements.push(value);
+        self
+    }
+
+    /// Decompose the batch
+    ///
+    /// Called by the user when they want to get at the internal store of
+    /// items. Returns a tuple, the first element being the allocated size of
+    /// stored items and the second the store of items.
+    fn destruct(self) -> (usize, Vec<I>) {
+        (self.allocated_bytes, self.elements)
+    }
+
+    /// Whether the batch has space for a new item
+    ///
+    /// This function returns true of there is space both in terms of item count
+    /// and byte count for the given item, false otherwise.
+    fn has_space(&self, value: &I) -> bool {
+        let too_many_elements = self.elements.len() + 1 > self.element_limit;
+        let too_many_bytes = self.allocated_bytes + value.size_of() > self.allocation_limit;
+        !(too_many_elements || too_many_bytes)
+    }
+
+    /// Push an element into the batch
+    ///
+    /// This function pushes an element into the batch. Callers must be sure to
+    /// call `has_space` prior to calling this function and receive a positive
+    /// result.
+    ///
+    /// # Panics
+    ///
+    /// This function will panic if there is not sufficient space in the batch
+    /// for a new element to be inserted.
+    fn push(&mut self, value: I) {
+        assert!(self.has_space(&value));
```

Comment:
Is this the right model here? This implies that safe usage of `push` involves calling `has_space` twice. Would it make sense to return something like `Result<(), I>` like some other `push` methods do?

Reply:
The `Result` approach would work. I mostly wanted to avoid needing to set bounds on `I` that require it to be debug for unwrap reasons. From the point of view of the caller you only need to call `has_space` the once. I would like to keep the panic in place to make it obvious that the function is not being called properly.

Reply:
Could do a `#[must_use]` with an `Option` return, or an equivalent enum like the batch buffer does.

Reply:
Sure. Now that we're talking about it one thing I do dislike about the check-the-return style is the calling code is much less straight line. Caller pushes, finds there's no space and does a flush. What do you do if you push again and there's still no space? This way the caller is guaranteed that their push will succeed. 

I'll leave this as-is unless there's consensus for fiddling with the return here.

Reply:
It feels like, with the way `Batcher` is written, that while this method in isolation looks weird, it's pretty straightforward in how it's used.  We might want to reconsider the design if we use this in multiple places, though.  Trying to sketch out the code in a check-the-return fashion yielded a lot of boilerplate that ultimately didn't feel much cleaner than this... so I'd be +1 of keeping it as-is.